feat(core): introduce Logger service

DzmitryShylovich opened the pull request:
Closes #5458

Commit:
feat(core): introduce Logger service

Closes #5458

FrozenPandaz commented:
Thoughts on implementing Logger.time as console.time?

https://developer.mozilla.org/en-US/docs/Web/API/Console/time

chunghha commented:
console.group as well please.

https://developer.mozilla.org/en-US/docs/Web/API/Console/group

mhevery commented:
I believe all JavaScript platforms have `console`. I that is the case we should move `Logger` into `core` and the default implementation should just do `console.log` 

What do you think?

gkalpak commented:
IE9 might need special treatment, because `console` is only available when DevTools are open (and has some quirks, like `console.log` not inheriting from `Function.prototype`).

DzmitryShylovich commented:
@gkalpak it's already handled [here](https://github.com/angular/angular/pull/14308/files#diff-f8c5c0ff8a5be8ae52ec4b1d4bd1da0cR41). Basically it's a function from [ng1](https://github.com/angular/angular.js/blob/master/src/ng/log.js#L138). If console is not available it will be noop.

gkalpak commented on modules/@angular/platform-browser/src/browser/logger/browser_console_logger.ts:
```diff
@@ -0,0 +1,73 @@
+  }
+}
+
+function formatError(arg: any): any {
```
We've been discussing about removing this method from AngularJS.
Stringifying the error, breaks source map supports (see https://github.com/angular/angular.js/issues/15590#issuecomment-278302504 for more info).

mgol replied:
Yeah, I'm in the process of removing this function for all the browsers with the exception of IE/Edge (as their error logging is far from ideal). Chrome/Safari apply source maps to logged errors, Firefox is in a process of implementing that as well. By adding this function you'll break it and if you use Webpack - which many people will as that's what Angular CLI uses - you'll just see references to `app.bundle.js` or `vendor.bundle.js` which makes making sense out of the stack virtually impossible.

DzmitryShylovich replied:
Currently this method is not used.

mgol replied:
That's true. Can it just be removed then?

DzmitryShylovich replied:
sure

DzmitryShylovich commented on modules/@angular/core/src/logger.ts:
```diff
@@ -0,0 +1,131 @@
+
+    // Note: reading logFn.apply throws an error in IE11 in IE8 document mode.
+    // The reason behind this is that console.log has type "object" in IE8...
+    try {
```
do we really need this check? afaik ng2 doesn't officially supported ie8

mgol replied:
Yeah, `try-catch` shouldn't be needed any more. Guarding against missing `console.log.apply` is as IE 9 has `console` method not inheriting from `Function.prototype`.

DzmitryShylovich replied:
removed

DzmitryShylovich commented:
@mhevery you're right, console is available on on supported platforms.
Moved `Logger` into `core`.

Also added `group` and `groupEnd` because these methods are used by the router.

mhevery commented:
rebase please

DzmitryShylovich commented:
@mhevery rebased
ci failed because of the invalid commit message in master 
```
Expected 'INVALID COMMIT MSG: "fix(Compiler): something"
```

mhevery commented:
We decided to not merge this in as at is a new feature and @IgorMinar and I need to debate whether we want to add and support this kind of features.

vicb commented on modules/@angular/core/src/logger.ts:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * Write a log message.
+   *
+   * @param args
```
Please remove `@param args` and make comment 1 liners

DzmitryShylovich replied:
done

vicb commented on modules/@angular/core/src/logger.ts:
```diff
@@ -0,0 +1,143 @@
+
+  groupEnd(): void { this._groupEnd(); }
+
+  private _getConsoleMethod(type: string): LogFn {
```
`private _getConsoleMethod(method, args)` and remove all the private wrappers.

DzmitryShylovich replied:
but in this case we will need to make unnecessary checks on every invocation.

DzmitryShylovich replied:
done

vicb commented on modules/@angular/core/testing/mock_logger.ts:
```diff
@@ -0,0 +1,69 @@
+ * @experimental
+ */
+export class MockLogger implements Logger {
+  /** Array of messages logged using info() */
```
Not sure if the comments are necessary on the Logs prop

DzmitryShylovich replied:
there's a similar comments in ng1. I thought it would be useful for generated api docs

DzmitryShylovich replied:
removed

vicb commented:
@mhevery @IgorMinar Yep, it would be good to have it, we use a mic of `console.log`, `Console` service today.

trumbitta commented:
I'm puzzled and can't figure out why a framework wouldn't have a logging system :-/

alastair-todd commented:
So, 2 months on 1 million people have introduced a nigh-on identical but independent "custom" LoggingService into their apps.

C'mon what's to consider here?

IgorMinar reviewed (changes requested):
My primary issue with this PR is that logging typically becomes very expensive (in terms of payload-size) very quickly. Until we have a design that can safely strip non-critical logs and minify the critical ones I think everyone is better off rolling their own solution. The api above would likely need to be changed to achieve these properties, and doing so after these apis are heavily used would be very disruptive and painful for everyone.

The cross-cutting concern of logging and necessity to provide a solution that works well for 3rd party components makes this a much harder problem than what it looks like on the surface and we are better off deferring tackling this problem rather than providing a  solution that would send everyone down the wrong path.

splincode commented:
@IgorMinar Should we close this task?

splincode commented:
@DzmitryShylovich Is the PR is relevant?

mhevery commented:
Closing per @IgorMinar's comment.
